src/back-end/lib/hooks/logger.ts

```diff
@@ -15,6 +24,19 @@ const hook: RouteHook<unknown, unknown, unknown, unknown, number, Session> = {
     request.logger.debug(
       `${chalk.gray("<-")} ${response.code} ${Date.now() - startTime}ms`
     );
+
+    if (LOG_MEM_USAGE) {
+      const heap = process.memoryUsage();
+      for (const key in heap) {
+        request.logger.debug(
+          `${key} ${
+            Math.round((heap[key as keyof MemoryUsage] / 1024 / 1024) * 100) /
+            100
+          } MB`
+        );
+      }
+      request.logger.debug("---");
+    }
```

Comment:
Each logger function (e.g. `logger.debug`) takes a third optional data argument. It's a plain object whose key/value pairs are printed to the console in a single line, delimited by spaces.

e.g.

```ts
logger.debug("domain", "msg", { foo: "bar", baz: 123 });
// [domain] msg foo=bar baz=123
```

It might help simplify the code and reduce log storage to use this feature — you'd only need to call `request.logger.debug("memory usage", heap)` once.

Reply:
Also, forgot to mention, my example is very simplified, I think the megabyte conversion you are doing is quite useful.

Reply:
So initially, I actually did measure heap usage on the `before` and `after` hooks to get the difference, but I realized that we don't necessarily have to do that. It results in two measurements instead of one, and there wasn't any measurable memory use occurring between requests in my testing. The heap usage at the end of the preceding request was always the exact amount measured at the start of the next request. So if we measure after each request, we're essentially getting the same measurement, with half the calls. I guess the potential disadvantage of doing it this way is that we could potentially miss any heap increases that happen in between requests, but that didn't seem to be happening. I'm happy to revert it back to the initial way I had if you think that's preferable.

I like the suggestion of generalizing using the existing optional parameter for the logging call. I'll make that change.

Reply:
Ah, sounds good to me, I don't think we need to make the two measurements. Thanks for the explanation.